Sources/Extraction/Filters/HillOrientation.c

```diff
@@ -1,26 +1,230 @@
+
+            for (int i = 0; i < NUM_VECTORS; i++)
+            {
+                Point neighbor = Calc_Add2Points(&neighbors[i], &(Point) { .x = x, .y = y });
+                Point antiPoint = (Point) { .x = -neighbors[i].x, .y = -neighbors[i].y };
+                Point antiNeighbor = Calc_Add2Points(&antiPoint, &(Point) { .x = x, .y = y });
+
+                //printf("considering neighbor (%d, %d)\n", neighbor.x, neighbor.y);
+
+                if(!IsInsideImageAndMask(imageHeight, imageWidth, pixelMask, neighbor.x, neighbor.y) || !IsInsideImageAndMask(imageHeight, imageWidth, pixelMask, antiNeighbor.x, antiNeighbor.y))
+                    continue;
+
+                float antiNeighborValue = input.data[antiNeighbor.x][antiNeighbor.y];
+                float pixelValue = input.data[x][y];
+                float neighborValue = input.data[neighbor.x][neighbor.y];
+
+                float strength = pixelValue - MAX(neighborValue, antiNeighborValue);
+
+                PointF orientation = Angle_ToVector(Angle_ToOrientation(Angle_Atan(neighbors[i])));
+                PointF contribution = Calc_Scalar_Multiply(strength, orientation);
+
+                /*printf("neighbor (%d, %d): %f\n", neighbor.x, neighbor.y, input.data[neighbor.x][neighbor.y]);
+                printf("anti-neighbor (%d, %d): %f\n", antiNeighbor.x, antiNeighbor.y, input.data[antiNeighbor.x][antiNeighbor.y]);
+                printf("strength: %f, orientation: (%f, %f), pixelValue: %f neighborValue: %f, anti-neighborValue: %f\n", strength, orientation.x, orientation.y, pixelValue, neighborValue, antiNeighborValue);
+                printf("Setting hill orientation for %d, %d: (%f, %f)\n", x, y, contribution.x, contribution.y);*/
+                
+                if (strength > 0) {
+                    directions.data[x][y] = Calc_Add2PointsF(&directions.data[x][y], &contribution);
+                }
+            }
+
+            //printf("FINISHED Calculate hill orientation for %d, %d: (%f, %f) ------------------------------------------\n", x, y, directions.data[x][y].x, directions.data[x][y].y);
+
+        }
+    }
+
+    //print_directions(directions);
+    return directions;
+}
+
+PointFArray2D HillOrientation_SumBlocks(const PointFArray2D * pixelOrientations, BinaryMap * blockMask, BlockMap * blocks)
 {
-    HillOrientation ho = {
-        .minHalfDistance = 2,    /* Lower = 0.5, Upper = 4 */
-        .maxHalfDistance = 6,    /* Lower = 5, Upper = 13 */
-        .neighborListSplit = 50, /* Upper = 100 */
-        .neighborsChecked = 20,  /* Upper = 100 */
-        .smoothingRadius = 1     /* Upper = 3 */
-    };
+    const PointFArray2D blockOrientations = PointFArray2D_Construct(blocks->blockCount.width, blocks->blockCount.height);
+    const int blockHeight = blocks->pixelCount.height / blocks->blockCount.height;
+    const int blockWidth = blocks->pixelCount.width / blocks->blockCount.width;
```

Comment:
Doesn't this information live in the block map already? Or is this deliberate recalculation to round down instead of up?